_posts/blog/2019-09-17-erasure-coding.md

```diff
@@ -0,0 +1,177 @@
+|gz | 2.7 | 2.7 | 21.3 |
+|none | 2.0 | 3.0 | 158.5 |
+|snappy | 1.6 | 1.6 | 38.4 |
+
+Of much more importance to Accumulo performance is read latency. A frequent use case for our group is to obtain a
+number of row IDs from an index and then use a BatchScanner to read those individual rows.
+In this use case, the time to access a single row is far more important that the raw I/O performance.  To test
+Accumulo's performance with EC for this use case, we did a series of tests against a 10 billion row table,
+with each row consisting of 10 columns.  16 Spark executors each performed 10000 queries, where each query
+sought 10 random rows.  Thus 16 million individual rows were returned in batches of 10.  For each batch of
+10, the time in milliseconds was captured, and theses times were collected in a histogram of 50ms buckets, with
+a catch-all bucket for queries that took over 1 second.  For this test we reconfigured our cluster to make use
+of the newer c5n.4xlarge nodes which feature must faster networking speeds (15 Gbps sustained vs 5 Gbps for 
+c5.4xlarge). Because these faster nodes are in short supply, we ran with only 16 HDFS nodes (c5n.4xlarge), 
+but still had 16 Spark nodes (also c5n.4xlarge).  Zookeeper and master nodes remained the same.
+
+In the table below, we show the min, max, and average times in milliseconds for each batch of 10 across
+four different encoding policies.  The clear winner here is replication, and the clear loser RS 10-4 with 
+1MB stripes, but RS 6-3 with 64KB stripes is not looking too bad.
+
+|Encoding|Min|Avg|Max|
+|--------|--:|--:|--:|
+|RS 10-4 1MB|40|105|2148|
+|RS 6-3 1MB|30|68|1297|
+|RS 6-3 64KB|23|43|1064|
+|Replication|11|23|731|
+
+The above results also hold in the event of errors.  The next table shows the same test, but with 2 DataNodes
+disabled to simulate system failures that require foreground rebuilds.  Again, replication wins, and RS 10-4 1MB
+loses, but RS 6-3 64KB remains a viable option.
+
+|Encoding|Min|Avg|Max|
+|--------|--:|--:|--:|
+|RS 10-4 1MB|53|143|3221|
+|RS 6-3 1MB|34|113|1662|
+|RS 6-3 64KB|24|61|1402|
+|Replication|12|26|304|
+
+The images below show a plots of the histograms.  The third plot was generated with 14 HDFS DataNodes, but after
+all missing data had been repaired.  Again, this was done to see how much of the performance degradation could be
+attributed to missing data, and how much to simply having less computing power available.
+
+<img src='/images/blog/201909_ec/ec-latency-16.png' width="75%">
+
+<img src='/images/blog/201909_ec/ec-latency-14e.png' width="75%">
+
+<img src='/images/blog/201909_ec/ec-latency-14.png' width="75%">
+
+### Conclusion
+HDFS with erasure coding has the potential to double your available Accumulo storage, at the cost of a hit in
+random seek times, but a potential increase in sequential scan performance. We hope to see Accumulo natively 
+support erasure coding at some point in the future.
```

Comment:
I don't understand this last sentence.  Are there changes that need to be made in Accumulo to better support EC?

Reply:
Ah, this is a shameless plug for the mods I talked about in the dev list post.  Can make this clearer or remove...what do you think?

Reply:
For a situation like this I would create an Accumulo issue or PR and have the blog post point to that.

Reply:
how about saying that we'll be proposing changes in the future to accommodate this.  I was kind of hoping for some reaction to my post on the dev list before creating an issue.  Should I just go ahead and create one anyway?

Reply:
added link to mailing list